src/core/lifecycle/prop-types.js

```diff
@@ -0,0 +1,117 @@
+import assert from 'assert';
+
+const TYPE_DEFINITIONS = {
+  boolean: {
+    validate(value, propType) {
+      return true;
+    }
+  },
+  number: {
+    validateType(value, propType) {
+      return (
+        'value' in propType &&
+        (!('max' in propType) || Number.isFinite(propType.max)) &&
+        (!('min' in propType) || Number.isFinite(propType.min))
+      );
+    },
+    validate(value, propType) {
+      return (
+        Number.isFinite(value) &&
+        (!('max' in propType) || value <= propType.max) &&
+        (!('min' in propType) || value >= propType.min)
+      );
+    }
+  }
+};
+
+export function parsePropTypes(propDefs) {
+  const propTypes = {};
+  const defaultProps = {};
+  for (const [propName, propDef] of Object.entries(propDefs)) {
+    const propType = parsePropType(propName, propDef);
+    propTypes[propName] = propType;
+    defaultProps[propName] = propType.value;
+  }
+  return {propTypes, defaultProps};
+}
+
+// Parses one property definition entry. Either contains:
+// * a valid prop type object ({type, ...})
+// * or just a default value, in which case type and name inference is used
+function parsePropType(name, propDef) {
+  switch (getTypeOf(propDef)) {
+    case 'object':
```

Comment:
Nit: use enum?

Reply:
>Nit: use enum?

Yes, I'd normally go with an enum, but here with one or two exceptions, these are the JavaScript built in `typeof` return values. Not clear to me that aliasing them with an enum makes it easier to read. 

The main reason is that in that case it feels like there should there be a mapping TO the enum as well in the `getTypeOf` function, which might be overkill.